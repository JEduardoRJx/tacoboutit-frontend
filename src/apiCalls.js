export const getRestaurants = async (lat = 39.7392, lng = -104.9903) => {
  const response = await fetch(`https://tacoboutit-backend.herokuapp.com/api/v1/restaurants/retrieve/?lat=${lat}&lng=${lng}`);
  if(!response.ok) {
    throw Error('Failed to fetch restaurants near you');
  }
  return response.json();
}

export const newTaco = async (type, restaurant) => {
  const response = await fetch('https://tacoboutit-backend.herokuapp.com/api/v1/tacos/new/', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ type, restaurant }),
  });
  if(!response.ok) {
    return Error('Failed to add taco.');
  }
  const resp = await response.json();
  return resp;
}

export const addReview = async (tacoId, rating, review = 'No Review Available') => {
  const response = await fetch(`https://tacoboutit-backend.herokuapp.com/api/v1/reviews/`, {
    method: 'POST', 
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      taco: tacoId,
      rating,
      review,
    }),
  })
  if(!response.ok) {
    return Error('Failed to add review');
  }
  const resp = await response.json();
  return resp;
}
