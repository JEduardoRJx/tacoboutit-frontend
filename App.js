import React, { Component } from "react";
import {
  StyleSheet,
  Text,
  View,
  FlatList,
  SafeAreaView,
  Image,
  Modal,
} from "react-native";
import { LinearGradient } from "expo-linear-gradient";
import { mockRestaurants } from "./mockRestaurants";
import { createAppContainer } from "react-navigation";
import { createBottomTabNavigator } from "react-navigation-tabs";
import RestaurantCard from './src/components/RestaurantCard/RestaurantCard';
import TacoSearch from './src/components/TacoSearch/TacoSearch';
import SplashPage from './src/components/SplashPage/SplashPage';
import { getRestaurants, newTaco } from './src/apiCalls';
import * as Location from 'expo-location';
import * as Permissions from 'expo-permissions';
import RestaurantPage from'./src/components/RestaurantPage/RestaurantPage'

export class App extends Component {
  state = {
    showModal: false,
    selectedRestaurant: null,
    restaurants: null,
    error: '',
    location: null,
    isLoading: true,
  };

  componentDidMount = async () => {
    try {
      const location = await this._getLocationAsync();
      const lat = location.coords.latitude;
      const lng = location.coords.longitude;
      const restaurants = await getRestaurants(lat, lng);
      restaurants.sort((a, b) => a.distance - b.distance)
      this.setState({ restaurants, isLoading: false });
      
    } catch {
      const restaurants = await getRestaurants();
      this.setState({ error: 'Failed to get tacos by location', restaurants, isLoading: false });
    }
  }

  _getLocationAsync = async () => {
    let { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== 'granted') {
      this.setState({
        error: 'Permission to access location was denied',
      });
    }

    let location = await Location.getCurrentPositionAsync({});
    this.setState({ location });
    return location;
  };

  handlePress = (id) => {
    const selectedRestaurant = this.state.restaurants.find((rest) => rest.id === id);
    this.setState({ showModal: true, selectedRestaurant });
  }

  submitNewTaco = async (type, restaurantId) => {
    const response = await newTaco(type, restaurantId);
    if(!response.error) {
      const newTaco = { type, restaurant: restaurantId };
      this.updateLocalTacos(newTaco);
    }
    return response;
  }

  updateLocalTacos = (newTaco) => {
    const localRestaurants = this.state.restaurants.map((rest) => rest);
    const restToUpdate = localRestaurants.find((rest) => rest.id === newTaco.restaurant);
    restToUpdate.tacos = [...restToUpdate.tacos, newTaco];
    this.setState({ restaurants: localRestaurants });
  }

  updateLocalReviews = (response) => {
    const newSelectedRestaurant = this.state.selectedRestaurant;
    const selectedTaco = newSelectedRestaurant.tacos.find(taco => taco.id === response.taco);
    selectedTaco.reviews.push(response)
    const restaurants = this.state.restaurants.map(rest => rest)
    let restaurantToUpdate = restaurants.find(rest => rest.id === newSelectedRestaurant.id);
    restaurantToUpdate = newSelectedRestaurant;
    this.setState({restaurants: restaurants})
  }
 
  render() {
    return (
      <LinearGradient
        colors={["#F0CB35", "#D56C2C", "#C02425"]}
        style={styles.container}
      >
        <TacoSearch />
        {this.state.restaurants && <FlatList data={this.state.restaurants} 
          showsVerticalScrollIndicator={false}
          renderItem={({ item }) => <RestaurantCard
          id={item.id}
          name={item.name}
          address={item.address}
          isClosed={item.is_closed}
          distance={item.distance}
          img={item.image_url}
          handlePress={this.handlePress}
          // key={item => item}
      />} />
      }
        <Modal 
          visible={this.state.showModal}
          animationType="fade"
          onRequestClose={() => {
            this.setState({ showModal: false, selectedRestaurant: null });
          }}
        >
          <RestaurantPage restaurant={this.state.selectedRestaurant} submitTaco={this.submitNewTaco} updateLocalReviews={this.updateLocalReviews}/>
        </Modal>
        {this.state.isLoading && <SplashPage isLoading={this.state.isLoading}/> }
      </LinearGradient>
    );
  }
}


class Tacos extends Component {
  render() {
    return (
      <View style={styles.container}>
        <Text>Taco Screen</Text>
      </View>
    );
  }
}

const TabNavigator = createBottomTabNavigator(
  {
    Home: {
      screen: App,
      navigationOptions: {
        tabBarIcon: () => (
          <Image
            source={require("./assets/home.png")}
            style={{ width: 40, height: 40 }}
          />
        )
      }
    },
    Tacos: {
      screen: Tacos,
      navigationOptions: {
        tabBarIcon: () => (
          <Image
            source={require("./assets/taco.png")}
            style={{ width: 40, height: 40 }}
          />
        )
      }
    }
  },
  {
    tabBarOptions: {
      showLabel: false,
      activeBackgroundColor: "rgba(0, 0, 0, .1)"
    }
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#ccc",
    alignItems: "center",
    justifyContent: "center"
  }
});

export default createAppContainer(TabNavigator);
