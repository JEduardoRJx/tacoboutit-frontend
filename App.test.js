import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import * as Permissions from 'expo-permissions';
import * as Location from 'expo-location';
import * as api from './src/apiCalls';

import { App } from './App';
import { mockRestaurants } from './mockRestaurants';

describe('App', () => {
  let wrapper;
  const mockNewTaco = jest.fn();
  const mockStateRestaurants = [
    {
        "id": 1,
        "yelp_id": "lAeWTnxwadpS20I25kSbnQ",
        "name": "Las Delicias North",
        "phone": "+13034300422",
        "is_closed": false,
        "review_count": 170,
        "yelp_rating": 3.5,
        "url": "https://www.yelp.com/biz/las-delicias-north-denver-2?adjust_creative=pxQ3XEqH9sM15FQYWsuBXQ&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=pxQ3XEqH9sM15FQYWsuBXQ",
        "latitude": 39.834828,
        "longitude": -104.986963,
        "image_url": "https://s3-media2.fl.yelpcdn.com/bphoto/p0JDroBv0kkPQ5eTenBcuQ/o.jpg",
        "address": "7610 Conifer Rd, Denver, CO 80221",
        "distance": 563.486550595786,
        "tacoboutit_item_review_count": 0,
        "tacos": [
            {
                "id": 9,
                "type": "Testing testing",
                "restaurant_id": 1,
                "average_rating": 5.0,
                "reviews": [
                    {
                        "id": 1,
                        "rating": 7.0,
                        "review": "its ok",
                        "taco": 9
                    },
                    {
                        "id": 2,
                        "rating": 4.0,
                        "review": "lame",
                        "taco": 9
                    },
                    {
                        "id": 3,
                        "rating": 4.0,
                        "review": null,
                        "taco": 9
                    },
                    {
                        "id": 4,
                        "rating": 1.0,
                        "review": "This taco put my son in the hospital",
                        "taco": 9
                    },
                    {
                        "id": 5,
                        "rating": 10.0,
                        "review": "The best",
                        "taco": 9
                    },
                    {
                        "id": 6,
                        "rating": 4.0,
                        "review": "meh",
                        "taco": 9
                    }
                ]
            }
        ]
    },
    {
        "id": 2,
        "yelp_id": "Oum4nOuOsyMRM5jnc6pGLQ",
        "name": "Freds Breakfast Burritos",
        "phone": "+13034283145",
        "is_closed": false,
        "review_count": 40,
        "yelp_rating": 5.0,
        "url": "https://www.yelp.com/biz/freds-breakfast-burritos-denver?adjust_creative=pxQ3XEqH9sM15FQYWsuBXQ&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=pxQ3XEqH9sM15FQYWsuBXQ",
        "latitude": 39.8234851,
        "longitude": -104.9892212,
        "image_url": "https://s3-media4.fl.yelpcdn.com/bphoto/BrjCYwtiLR-R5qyddrnjnw/o.jpg",
        "address": "130 W 70th Ave, Denver, CO 80221",
        "distance": 1782.59977835185,
        "tacoboutit_item_review_count": 0,
        "tacos": []
    }];

  beforeEach(() => {
     wrapper = shallow(<App />);
     wrapper.setState({ restaurants: mockStateRestaurants});
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to handlePress', async () => {
    expect(wrapper.state('restaurants').length).toBe(2);
    await wrapper.instance().handlePress(1);
    expect(wrapper.state('selectedRestaurant')).toEqual(mockStateRestaurants[0]);
  })

  it('submitNewTaco should return the response from BE', async () => {
    const mockResponse = { 
      id: 1, 
      restaurantId: 2, 
      averageRating: null, 
      type: 'carne asada', 
      reviews: [],
    };
    api.newTaco = jest.fn().mockImplementation(() => Promise.resolve(mockResponse));
    const mockUpdateLocalTacos = jest.fn();
    wrapper.instance().updateLocalTacos = mockUpdateLocalTacos;
    const result = await wrapper.instance().submitNewTaco('carne asada', 2);
    expect(mockUpdateLocalTacos).toHaveBeenCalledWith({"restaurant": 2, "type": "carne asada"});
    expect(result).toEqual(mockResponse);
  });

  it('updateLocalTacos should add a taco to a restaurant in state', () => {
    const mockNewTaco = { restaurant: 1, type: 'test type', id: 23, reviews: [] };
    wrapper.instance().updateLocalTacos(mockNewTaco);
    expect(wrapper.state('restaurants')[0].tacos[1]).toEqual(mockNewTaco);
  });
});