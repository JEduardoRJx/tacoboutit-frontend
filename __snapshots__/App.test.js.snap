// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<LinearGradient
  colors={
    Array [
      "#F0CB35",
      "#D56C2C",
      "#C02425",
    ]
  }
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ccc",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <FlatList
    data={
      Array [
        Object {
          "address": "2835 W 72nd Ave, Westminster, CO 80030",
          "id": 1,
          "name": "Guadalajara Mexican Restaurant",
        },
        Object {
          "address": "1111 W Taco St, CO 80211",
          "id": 2,
          "name": "Taco Taco Taco",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <Component
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <RestaurantPage
      restaurant={null}
      submitTaco={[Function]}
      updateLocalReviews={[Function]}
    />
  </Component>
  <SplashPage
    isLoading={true}
  />
</LinearGradient>
`;
